storage/filestore/dir.go

```diff
@@ -238,6 +348,136 @@ func (dir *Dir) GarbageCollect(ctx context.Context) (err error) {
 	return nil
 }
 
+const nameBatchSize = 1024
+
+// GetAllNamespaces finds all known namespace IDs in use in local storage. They are not
+// guaranteed to contain any blobs.
+func (dir *Dir) GetAllNamespaces(ctx context.Context) (ids [][]byte, err error) {
+	defer mon.Task()(&ctx)(&err)
+	topBlobDir := dir.blobsdir()
+	openDir, err := os.Open(topBlobDir)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { err = errs.Combine(err, openDir.Close()) }()
+	for {
+		dirNames, err := openDir.Readdirnames(nameBatchSize)
+		if err != nil && err != io.EOF {
+			return nil, err
+		}
+		if len(dirNames) == 0 {
+			return ids, nil
+		}
+		for _, name := range dirNames {
+			namespace, err := pathEncoding.DecodeString(name)
+			if err != nil {
+				// just an invalid directory entry, and not a namespace. probably
+				// don't need to pass on this error
+				continue
+			}
+			ids = append(ids, namespace)
+		}
+	}
+}
+
+// ForAllKeysInNamespace executes doForEach for each locally stored blob, stored with storage
+// format V1 or greater, in the given namespace. If doForEach returns a non-nil error,
+// ForAllKeysInNamespace will stop iterating and return the error immediately.
+func (dir *Dir) ForAllKeysInNamespace(ctx context.Context, namespace []byte, doForEach func(storage.StoredBlobAccess) error) (err error) {
```

Comment:
It looks like this is all so sequentially. Is there a way to gain performance benefits by adding some concurrency, either here or in `forAllKeysInNamespaceWithPrefix`? It seems like with an errGroup (or similar) you could easily gather all errors and could make the iteration much faster.

Reply:
That would not be a good fit for all potential uses of this visitor pattern, I don't think. For example, in all the existing uses, I believe we would rather slow it down than speed it up (we go as far as to call `runtime.GoSched()` in the callback in some cases, to ensure that this goroutine does not use up too much cpu time at the expense of more important things). In cases where a certain task does make sense to do concurrently, the callback can invoke goroutines as desired. If it needs to get even more performance squeezed out, it might make sense to parallelize the second-tier directory iterations, but I am not sure that would help much, and it would be a lot harder to think about in terms of its relationship to the rest of the program.